pkg/cache/cluster.go

```diff
@@ -615,6 +630,29 @@ func (c *clusterCache) watchEvents(ctx context.Context, api kube.APIResourceInfo
 				if errors.IsNotFound(err) {
 					c.stopWatching(api.GroupKind, ns)
 				}
+				var connectionUpdated bool
+				if err != nil {
```

Comment:
Hey Chetan, me again, a few more thoughts while thinking about the ramifications of the cluster invalidation `c.Invalidate()` and `c.EnsureSynced()` calls. (Feel free to poke holes in my logic!)

- The watch function (defined here) is called for each watched resource, so if there are 100 resources, it will be called once for each, to establish the watch.
- What about the scenario where only 1 watch (or a small number are failing), for example because the ServiceAccount does not have the correct permissions to watch the resource?
	- If one watch fails 1 out of 100, will the cluster oscillate between Successful and Failed?
	- It might look like this:
	- Watch 1: succeeds, set cluster status to Successful
	- Watch 2: succeeds, set cluster status to Successful
	- Watch 3: succeeds, set cluster status to Successful
	- (...)
	- Watch 80: succeeds, set cluster status to Failed
	- Perhaps the ideal behaviour would be to report success/failure only after all were established. (Easier said than done, I'm sure)
	
- Another question: since it appears we are invalidating the cluster cache on any failure:
	- Will this cause cache invalidation and re-sync to occur much more often, if the network connection between Argo CD and the target (watched) cluster happens to be unstable? (for example, imagine a case where 1% of connections fail. Would this magnify to invalidating and forcing a resync of all the resources, if only 1% of connections were failing?).

Reply:
Hi Jonathan, your concerns are valid. Since watches could fail for any reason it is difficult to determine the cluster status based on the errors returned during watch. These errors could be transient making it complicated to keep track of their states. Also, the error message could be generic to determine why exactly the watch failed.

To fix this issue, I have included a goroutine that periodically checks for watch errors, pings the remote cluster(to get the version), and updates the status. By following this approach we don't rely completely on the watches and prevent the status from oscillating in the case of transient watch failures. We use the watch status to prevent the number of pings to the remote clusters by pinging the cluster only when the watches fail. I've also added checks to ensure we are invalidating the cluster cache only when the status changes which shouldn't happen frequently. 

Let me know what you think. The periodic goroutine approach does handle some of the edge cases that may arise due to the dynamic nature of the watches. But also I'm open to hearing other approaches to solve this problem.